### PR TITLE
Runtime support for the NSArchiver class attributes.

### PR DESCRIPTION
**Pull request by eeckstein**
Register class names for NSKeyedArchiver and NSKeyedUnarchiver based on the @NSKeyedArchiveLegacy and @_staticInitializeObjCMetadata class attributes.

@NSKeyedArchiveLegacy registers a class name translation.
@_staticInitializeObjCMetadata just makes sure that the metadata of a class is instantiated.

This registration code is executed as a static initializer, like a C++ global constructor.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

**Comment by eeckstein** [comment]
@DougGregor, @jrose-apple, @parkera Can you please take a look?

**Comment by eeckstein** [comment]
@swift-ci Please smoke test

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 1]
```diff
@@ -95,3 +95,20 @@ extension CVarArg where Self: _ObjectiveCBridgeable {
...
+public func _registerClassNameForArchiving(_ NameForClass: UnsafePointer<CChar>,
+                                           _ Class: AnyClass) {
+  // If it's not possible to create a String from the name, it should abort
+  // and not fail silently.
+  let NameStr = String(validatingUTF8: NameForClass)!
+
+  // Register the class name mapping for archiving and unarchiving.
+  NSKeyedArchiver.setClassName(NameStr, for: Class)
+  NSKeyedUnarchiver.setClass(Class, forClassName: NameStr)
+}
```
Comment (parkera) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Nitpick, but this looks like the llvm naming conventions translated into Swift. This should have lowercase variable names.

**Comment by parkera** [comment]
also, @itaiferber

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 2]
```diff
@@ -95,3 +95,20 @@ extension CVarArg where Self: _ObjectiveCBridgeable {
...
+//===----------------------------------------------------------------------===//
+// Runtime support for NSKeyedArchives
+//===----------------------------------------------------------------------===//
+
+@_silgen_name("swift_registerClassNameForArchiving")
+public func _registerClassNameForArchiving(_ NameForClass: UnsafePointer<CChar>,
+                                           _ Class: AnyClass) {
+  // If it's not possible to create a String from the name, it should abort
+  // and not fail silently.
+  let NameStr = String(validatingUTF8: NameForClass)!
```
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/Foundation.swift]: We do want this to be as little of a performance hit as possible, though. If there's a non-validating version I would prefer that.

**Comment by eeckstein** [comment]
@swift-ci Please smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**